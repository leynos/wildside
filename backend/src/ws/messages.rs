//! WebSocket message types for user events.
use actix::Message;
use serde::Serialize;
use uuid::Uuid;

/// Event emitted when a new user is created.
#[derive(Debug, Serialize, Message)]
#[rtype(result = "()")]
pub struct UserCreated {
    #[serde(rename = "trace_id")]
    pub trace_id: String,
    pub id: String,
    pub display_name: String,
}

impl UserCreated {
    /// Construct with a fresh trace identifier.
    pub fn new(id: impl Into<String>, display_name: impl Into<String>) -> Self {
        Self {
            trace_id: Uuid::new_v4().to_string(),
            id: id.into(),
            display_name: display_name.into(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;
    use serde_json::Value;

    #[rstest]
    fn serializes_user_created() {
        let event = UserCreated::new("123", "Alice");
        let value = serde_json::to_value(&event).unwrap();
        assert!(value.get("trace_id").is_some());
        assert_eq!(value.get("id").and_then(Value::as_str), Some("123"));
        assert_eq!(
            value.get("display_name").and_then(Value::as_str),
            Some("Alice")
        );
        insta::assert_json_snapshot!(value, {
            ".trace_id" => "[trace_id]"
        });
    }
}
