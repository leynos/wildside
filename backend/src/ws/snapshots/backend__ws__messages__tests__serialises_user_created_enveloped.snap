---
source: src/ws/messages.rs
assertion_line: 94
expression: value
---
{
  "displayName": "Alice",
  "id": "123",
  "traceId": "00000000-0000-0000-0000-000000000000"
}
